src/utils/graphql/smartweave.ts

```diff
@@ -0,0 +1,389 @@
+ */
+export async function getSmartweaveContractsFromGQL({
+  address,
+  arweave,
+  signal,
+  // TODO: add retry provider - eg pass in fetcher for the query
+}: {
+  arweave: Arweave;
+  address: string;
+  signal?: AbortSignal;
+}): Promise<DataProtocolTransaction[]> {
+  let hasNextPage = false;
+  let cursor: string | undefined;
+  const protocolTxs = new Set<DataProtocolTransaction>();
+  do {
+    const queryObject = {
+      query: `
+      { 
+          transactions (
+              owners:["${address}"]
+              tags: [
+                      {
+                        name: 'App-Name',
+                        values: ['SmartWeaveContract'],
+                      },
+                    ],
+              sort: HEIGHT_DESC,
+              first: ${MAX_REQUEST_SIZE},
+              bundledIn: null,
+              ${cursor !== undefined ? `after: "${cursor}"` : ''}
+          ) {
+              pageInfo {z
+                  hasNextPage
+              }
+              edges {
+                  cursor
+                  node {
+                      id
+                      block {
+                          height
+                      }
+                      tags {
+                          name
+                          value
+                      }
+                  }
+              }
+          }
+      }`,
+    };
+
+    if (signal?.aborted) {
+      throw new AbortError('GraphQL request was aborted.');
+    }
+
+    const { status, ...response } = await arweave.api.post(
+      '/graphql',
+      queryObject,
+      {
+        signal,
+      },
+    );
+
+    if (status == 429) {
+      await new Promise((resolve) => setTimeout(resolve, 10_000));
+      continue;
+    }
+
+    if (status !== 200) {
+      throw Error(
+        `Failed to fetch contracts for wallet. Status code: ${status}`,
+      );
+    }
+
+    if (!response.data.data?.transactions?.edges?.length) {
+      continue;
+    }
+    response.data.data.transactions.edges
+      .map((e: { node: GQLNodeInterface }) => ({
+        id: e.node.id,
+        tags: e.node.tags,
+        data: e.node.data,
+      }))
+      .forEach((c: { def: DataProtocolTransaction; cursor: string }) => {
+        protocolTxs.add(c.def);
+      });
+    cursor =
+      response.data.data.transactions.edges[MAX_REQUEST_SIZE - 1]?.cursor;
+    hasNextPage =
+      response.data.data.transactions.pageInfo?.hasNextPage ?? false;
+  } while (hasNextPage);
+
+  return [...protocolTxs];
+}
+
+/**
+ * @param arweave - arweave instance to perform gql request
+ * @param address - address of the wallet to get transactions for
+ * @param contractTxId - contract tx id to filter the transactions
+ * @param signal - signal to abort the request
+ * @returns @type {Promise<DataProtocolTransaction[]>} - returns the list of data protocol txs, for e.g smartweave contracts
+ */
+export async function getSmartweaveTransactionsFromGQL({
+  address,
+  contractTxId,
+  arweave,
+  signal,
```

Comment:
we may want to hard code this to be like 10 seconds or something (e.g `signal = AbortSignal.timeout(10_000)`)

Reply:
https://github.com/ar-io/ar-io-sdk/pull/104/commits/6ef3aa5b3553ed4c67524f56a6ff223fd60dc0a8